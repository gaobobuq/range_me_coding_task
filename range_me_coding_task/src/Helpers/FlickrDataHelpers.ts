class FlickrDataHelpers {
  public static extractAuthorName = (authorString?: string): string => {
    return !!authorString
      ? authorString.substring(authorString.indexOf('("') + 2, authorString.lastIndexOf('")')).trim()
      : '';
  };

  public static extractAuthorEmail = (authorString?: string): string => {
    return !!authorString
      ? authorString.substring(0, authorString.lastIndexOf('("')).trim()
      : '';
  };
}

export default FlickrDataHelpers;
