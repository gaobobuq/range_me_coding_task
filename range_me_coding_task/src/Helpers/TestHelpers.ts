import { ReactWrapper } from 'enzyme';

/**
 * TestHelpers
 */
class TestHelpers {
  /**
   * checkTextValue
   * check whether the text value is identical as expected
   * @param wrapper enzyme react wrapper
   * @param selector target text selector
   * @param expectedValue the expected value
   */
  public static checkTextValue = (wrapper: ReactWrapper, selector: string, expectedValue: string): void => {
    const textElement = wrapper.find(selector);
    expect(textElement.text()).toBe(expectedValue);
  };

  /**
   * printWrapperContent
   * print debug wrapper content for debugging only
   * @param wrapper the input wrapper
   */
  public static printWrapperContent = (wrapper: ReactWrapper): void => {
    console.log(wrapper.debug());
  };
}

export default TestHelpers;
