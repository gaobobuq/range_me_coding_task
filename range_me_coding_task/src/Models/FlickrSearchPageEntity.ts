import { JsonFlickrFeedEntity, QueryParametersEntity } from '.';

export default class FlickrSearchPageEntity {
  public queryParameters?: QueryParametersEntity;
  public flickrFeed?: JsonFlickrFeedEntity;
}
