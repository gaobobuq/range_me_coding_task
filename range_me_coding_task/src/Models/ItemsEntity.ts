import { MediaEntity } from '.';
import { ItemsRawData } from './RawData';

export default class ItemsEntity {
  public title?: string;
  public link?: string;
  public media?: MediaEntity;
  public dateTaken?: Date;
  public description?: string;
  public published?: Date;
  public author?: string;
  public authorId?: string;
  public tags?: string;

  constructor(rawData: ItemsRawData) {
    this.title = rawData.title;
    this.link = rawData.link;

    if (rawData.media) {
      this.media = new MediaEntity(rawData.media);
    }

    if (rawData.date_taken) {
      this.dateTaken = new Date(rawData.date_taken);
    }

    this.description = rawData.description;

    if (rawData.published) {
      this.published = new Date(rawData.published);
    }

    this.author = rawData.author;
    this.authorId = rawData.author_id;
    this.tags = rawData.tags;
  }
}
