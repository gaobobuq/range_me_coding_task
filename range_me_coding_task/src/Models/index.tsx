export { default as FlickrSearchPageEntity } from './FlickrSearchPageEntity';
export { default as QueryParametersEntity } from './QueryParametersEntity';
export { default as MediaEntity } from './MediaEntity';
export { default as ItemsEntity } from './ItemsEntity';
export { default as JsonFlickrFeedEntity } from './JsonFlickrFeedEntity';
