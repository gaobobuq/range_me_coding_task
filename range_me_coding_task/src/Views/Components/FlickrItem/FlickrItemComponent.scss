
.data-item {
  margin: 20px;
  padding: 20px;
  min-width: 160px;

  .media-image {
    padding: 0 15px;
    width: 100%;
    max-height: 208px;
    object-fit: contain;
    box-shadow: 0 3px 3px rgba(0, 0, 0, 0.2);
  }

  .author {
    font-size: 18px;
    border-style: solid;
    border-width: 0 0 2px 0;
    border-color: #d7d7d7;
    font-weight: normal;
    color: #858585;
    margin-bottom: 12px;
  }

  .tags {
    color: #2138ac;
  }
}
