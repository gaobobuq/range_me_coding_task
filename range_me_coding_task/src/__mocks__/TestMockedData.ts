import { ItemsRawData, JsonFlickrFeedRawData } from '../Models/RawData';

export const itemsEntityRawData: ItemsRawData = {
  title: 'IMG_2481',
  link: 'https://www.flickr.com/photos/mickythepixel/51893643867/',
  media: {
    m: 'https://live.staticflickr.com/65535/51893643867_051d38f43e_m.jpg',
  },
  date_taken: '2022-02-19T12:08:25-08:00',
  description:
      ' <p><a href="https://www.flickr.com/people/mickythepixel/">micky the pixel</a> posted a photo:</p> <p><a href="https://www.flickr.com/photos/mickythepixel/51893643867/" title="IMG_2481"><img src="https://live.staticflickr.com/65535/51893643867_051d38f43e_m.jpg" width="230" height="240" alt="IMG_2481" /></a></p> ',
  published: '2022-02-21T09:23:45Z',
  author: 'nobody@flickr.com ("micky the pixel")',
  author_id: '7141646@N02',
  tags: 'buch book livre zeitschrift magazine verlagernstkeilsnachfolger diegartenlaubeâ€“illustriertesfamilienblatt diegartenlaube illustration wilhelmvonkaulbach reinekefuchs hinrichtung execution',
};

export const jsonFlickrFeedRawDataOneInstance: JsonFlickrFeedRawData = {
  title: 'Uploads from everyone',
  link: 'https://www.flickr.com/photos/',
  description: '',
  modified: '2022-02-21T09:23:41Z',
  generator: 'https://www.flickr.com',
  items: [
    {
      title: ' ',
      link: 'https://www.flickr.com/photos/192801292@N06/51893643787/',
      media: {
        m: 'https://live.staticflickr.com/65535/51893643787_c2269ca5d9_m.jpg',
      },
      date_taken: '2022-02-21T14:27:58-08:00',
      description:
                ' <p><a href="https://www.flickr.com/people/192801292@N06/">abdulrazacksa</a> posted a photo:</p> <p><a href="https://www.flickr.com/photos/192801292@N06/51893643787/" title=" "><img src="https://live.staticflickr.com/65535/51893643787_c2269ca5d9_m.jpg" width="240" height="195" alt=" " /></a></p> ',
      published: '2022-02-21T09:23:41Z',
      author: 'nobody@flickr.com ("abdulrazacksa")',
      author_id: '192801292@N06',
      tags: '',
    },
  ],
};

export const jsonFlickrFeedRawDataNoItems: JsonFlickrFeedRawData = {
  title: 'Uploads from everyone',
  link: 'https://www.flickr.com/photos/',
  description: '',
  modified: '2022-02-21T09:23:41Z',
  generator: 'https://www.flickr.com',
  items: [],
};
