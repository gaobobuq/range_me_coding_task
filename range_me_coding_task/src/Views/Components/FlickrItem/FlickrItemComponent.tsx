import * as React from 'react';
import { Col, Row } from 'react-grid-system';
import { ItemsEntity } from '../../../Models';
import FlickrDataHelpers from '../../../Helpers/FlickrDataHelpers';

import '../../frontend.scss';
import './FlickrItemComponent.scss';

/**
 * FlickrItemComponent Component
 * Component to display single Flickr item
 * @param author author information
 * @param dateTaken time taken for the image
 * @param link  link to the full image
 * @param media thumbnail image
 * @param tags data item tags
 */
function FlickrItemComponent({
  author, dateTaken, link, media, tags,
}: ItemsEntity) {
  return (
    <div className="data-item paper text-base-style">
      <Row>
        <Col md={4} sm={12}>
          {/* image thumbnail */}

          <img
            className="media-image"
            src={media?.m}
            alt={author}
            onError={(e: any) => { e.target.src = 'no-image-placeholder.png'; }}
          />
        </Col>

        <Col md={8} sm={12}>
          {/* author name */}
          <div className="paragraph author" data-testid="author-name">{FlickrDataHelpers.extractAuthorName(author)}</div>

          {/* author email */}
          <div className="paragraph" data-testid="author-email">{FlickrDataHelpers.extractAuthorEmail(author)}</div>

          {/* date taken */}
          <div className="label" data-testid="date-taken">{dateTaken?.toUTCString()}</div>

          {/* tags */}
          <div className="tags" data-testid="tags">{tags}</div>

          {/* link to full image */}
          <div>
            <a className="link" data-testid="link-to-full-image" href={link}>
              View Full Image
            </a>
          </div>
        </Col>
      </Row>
    </div>
  );
}

export default FlickrItemComponent;
