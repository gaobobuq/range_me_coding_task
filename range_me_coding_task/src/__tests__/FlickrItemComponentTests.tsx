import { mount } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import FlickrDataHelpers from '../Helpers/FlickrDataHelpers';
import TestHelpers from '../Helpers/TestHelpers';
import { ItemsEntity } from '../Models';
import FlickrItemComponent from '../Views/Components/FlickrItem/FlickrItemComponent';
import { itemsEntityRawData } from '../__mocks__/TestMockedData';

/**
 * Flickr Item Component Tests
 */
describe('(Tags: RMCT-004 FlickrItem) Flickr Item Component Tests', () => {
  it('FlickrItem component can be rendered', () => {
    const itemsEntity = new ItemsEntity(itemsEntityRawData);

    const wrapper = mount(
      <FlickrItemComponent {...itemsEntity} />,
    );

    expect(wrapper).not.toBeNull();
  });

  /**
   * Test scenarios
   */
  it('FlickrItem component elements have expected values', () => {
    const itemsEntity = new ItemsEntity(itemsEntityRawData);

    const wrapper = mount(
      <FlickrItemComponent {...itemsEntity} />,
    );

    // check author name value
    TestHelpers.checkTextValue(wrapper, 'div[data-testid="author-name"]', FlickrDataHelpers.extractAuthorName(itemsEntity.author));

    // check author email value
    TestHelpers.checkTextValue(wrapper, 'div[data-testid="author-email"]', FlickrDataHelpers.extractAuthorEmail(itemsEntity.author));

    // check date taken value
    TestHelpers.checkTextValue(wrapper, 'div[data-testid="date-taken"]', itemsEntity.dateTaken.toUTCString());

    // check tags value
    TestHelpers.checkTextValue(wrapper, 'div[data-testid="tags"]', itemsEntity.tags);

    // check full image link
    const linkElement = wrapper.find('a[data-testid="link-to-full-image"]');
    expect(linkElement.prop('href')).toBe(itemsEntity.link);

    expect(wrapper).not.toBeNull();
  });

  it('Check FlickrItem component details using snapshot', () => {
    const itemsEntity = new ItemsEntity(itemsEntityRawData);

    const renderDetails = renderer
      .create(<FlickrItemComponent {...itemsEntity} />)
      .toJSON();

    expect(renderDetails).toMatchSnapshot();
  });
});
