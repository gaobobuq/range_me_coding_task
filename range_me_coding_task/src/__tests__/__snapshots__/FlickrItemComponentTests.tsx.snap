// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Tags: RMCT-004 FlickrItem) Flickr Item Component Tests Check FlickrItem component details using snapshot 1`] = `
<div
  className="data-item paper text-base-style"
>
  <div
    style={
      Object {
        "alignItems": "normal",
        "display": "flex",
        "flexDirection": "row",
        "flexGrow": 0,
        "flexShrink": 0,
        "flexWrap": "wrap",
        "justifyContent": "flex-start",
        "marginLeft": -15,
        "marginRight": -15,
      }
    }
  >
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "flexBasis": "33.333333333333336%",
          "flexGrow": 0,
          "flexShrink": 0,
          "left": "auto",
          "marginLeft": "0%",
          "maxWidth": "33.333333333333336%",
          "minHeight": 1,
          "paddingLeft": 15,
          "paddingRight": 15,
          "position": "relative",
          "right": "auto",
          "width": "33.333333333333336%",
        }
      }
    >
      <img
        alt="nobody@flickr.com (\\"micky the pixel\\")"
        className="media-image"
        onError={[Function]}
        src="https://live.staticflickr.com/65535/51893643867_051d38f43e_m.jpg"
      />
    </div>
    <div
      style={
        Object {
          "boxSizing": "border-box",
          "flexBasis": "66.66666666666667%",
          "flexGrow": 0,
          "flexShrink": 0,
          "left": "auto",
          "marginLeft": "0%",
          "maxWidth": "66.66666666666667%",
          "minHeight": 1,
          "paddingLeft": 15,
          "paddingRight": 15,
          "position": "relative",
          "right": "auto",
          "width": "66.66666666666667%",
        }
      }
    >
      <div
        className="paragraph author"
        data-testid="author-name"
      >
        micky the pixel
      </div>
      <div
        className="paragraph"
        data-testid="author-email"
      >
        nobody@flickr.com
      </div>
      <div
        className="label"
        data-testid="date-taken"
      >
        Sat, 19 Feb 2022 20:08:25 GMT
      </div>
      <div
        className="tags"
        data-testid="tags"
      >
        buch book livre zeitschrift magazine verlagernstkeilsnachfolger diegartenlaubeâ€“illustriertesfamilienblatt diegartenlaube illustration wilhelmvonkaulbach reinekefuchs hinrichtung execution
      </div>
      <div>
        <a
          className="link"
          data-testid="link-to-full-image"
          href="https://www.flickr.com/photos/mickythepixel/51893643867/"
        >
          View Full Image
        </a>
      </div>
    </div>
  </div>
</div>
`;
